src/topology/mod.rs

```diff
@@ -689,13 +718,17 @@ impl TopologyDescription {
             }
         }
 
-        self.add_missing_hosts(&description, client, top_arc, run_monitor);
+        self.add_missing_hosts(description.clone(), client, top_arc, run_monitor);
 
         // Remove hosts that are not reported by the primary.
         let mut hosts_to_remove = Vec::new();
         for host in self.servers.keys() {
-            if !description.hosts.contains(host) && !description.passives.contains(host) &&
-               !description.arbiters.contains(host) {
+            let description_guard = description.read().unwrap();
+
+            if !description_guard.hosts.contains(host) &&
+                !description_guard.passives.contains(host) &&
+                !description_guard.arbiters.contains(host)
+            {
```

Comment:
With this kind of thing, I tend to be of the opinion that the only way to really determine what performs better is to actually run benchmarks rather than abstractly reasoning about it. Unfortunately, we still haven't had a chance to implement any benchmarks for the driver (although I think we have an issue open for it), so it's probably best to leave the behavior as possible close to as what it was before, which would be acquiring the lock only once as you suggest. I'll implement your suggestion before merging.